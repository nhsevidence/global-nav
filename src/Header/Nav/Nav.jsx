import React, { Component } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import SubNav from "./SubNav";
import styles from "./Nav.module.scss";
import rootLinks from "./links.json";

export default class Nav extends Component {
	render() {
		const { accountsLinks } = this.props;

		// Links from NICE Accounts is an object so flatten to a loop for easier traversal
		const accountsLinksArray =
			accountsLinks &&
			Object.keys(accountsLinks).map(text => ({
				text: text,
				href: accountsLinks[text]
			}));

		const activeService = rootLinks.find(
			function(link) {
				return this.props.service && link.id === this.props.service;
			}.bind(this)
		);

		const subLinks = activeService && activeService.links;

		return (
			<div
				id="header-menu"
				className={classnames(
					styles.wrapper,
					{
						[styles.wrapperExpanded]: this.props.isExpanded
					},
					{
						[styles.wrapperWithSubLinks]: subLinks
					}
				)}
			>
				<nav className={styles.nav}>
					<div className={styles.menuWrapper}>
						<ul
							className={styles.menuList}
							role="menu"
							aria-labelledby="header-menu-button"
						>
							{rootLinks.map(({ href, id, text, abbreviation, title }) => {
								let ariaCurrent = null;

								if (this.props.service && id === this.props.service) {
									ariaCurrent = true;

									if (
										location &&
										href ===
											`${location.protocol}//${location.host}${
												location.pathname
											}`
									) {
										ariaCurrent = "page";
									}
								}

								return (
									<li key={id} role="presentation">
										<a
											href={href}
											aria-current={ariaCurrent}
											role="menuitem"
											className={styles.link}
										>
											<span>
												{abbreviation ? (
													<abbr title={title}>{text}</abbr>
												) : (
													text
												)}
											</span>
										</a>
										{ariaCurrent && subLinks && (
											<SubNav links={subLinks} text={text} />
										)}
									</li>
								);
							})}
						</ul>
					</div>
				</nav>
				{accountsLinksArray && (
					<nav
						aria-label="My account"
						className={classnames(styles.nav, styles.myAccount)}
					>
						{accountsLinksArray.length > 1 && (
							<h2 className={styles.myAccountHeading}>My account</h2>
						)}
						<div className={styles.menuWrapper}>
							<ul className={styles.menuList} role="menu">
								{accountsLinksArray.map(({ href, text }) => (
									<li key={href} role="presentation">
										<a href={href} role="menuitem" className={styles.link}>
											{text}
										</a>
									</li>
								))}
							</ul>
						</div>
					</nav>
				)}
			</div>
		);
	}
}

Nav.propTypes = {
	service: PropTypes.string,
	isExpanded: PropTypes.bool,
	accountsLinks: PropTypes.object
};
