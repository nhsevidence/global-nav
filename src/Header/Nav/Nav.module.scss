@import '~@nice-digital/design-system/src/stylesheets/helpers/helpers-utils';
@import 'mq';

$colour-nav-background: $colour-nice-teal;

.nav {
  background: $colour-nav-background;
  display: none;
  padding: rem($spacing-small 0);

  a {
    color: $colour-text-inverse;
    display: block;
    padding: rem($spacing-small $spacing-medium);
    text-decoration: none;

    &[aria-current='true'],
    &[aria-current='page'] {
      background: $colour-nice-white; // TODO: Update to $colour-body when the Design System is used everywhere
      color: $colour-text;
      font-weight: bold;
    }

    &:focus {
      background: $colour-nice-purple;
      color: $colour-text-inverse;
      outline: rem($spacing-x-small) solid $colour-nice-purple;
      z-index: 1; // Push above the adjacent nav items
    }

    &:hover {
      background: $colour-nice-teal-2;
      color: $colour-text-inverse;
    }
  }

  abbr {
    text-decoration: none;
  }

  @include mq($from: md) {
    display: block;
    padding: 0;

    a {
      height: 100%;
      padding: rem($spacing-medium $spacing-medium);
      position: relative;
      text-align: center;
      white-space: nowrap;

      span {
        display: inline-block;
        vertical-align: middle;
        white-space: normal;
      }

      // Use pseudo to centrally align text
      &:before {
        content: '';
        display: inline-block;
        height: 100%;
        vertical-align: middle;
      }

      &[aria-current='true'],
      &[aria-current='page'] {
        &:after {
          background: $colour-nav-background;
          content: '';
          display: block;
          height: rem($spacing-x-small);
          left: 0;
          position: absolute;
          top: 0;
          width: 100%;
        }
      }
    }
  }
}

.navExpanded {
  display: block;
}

.menu {
  list-style: none;
  margin: auto;
  padding: 0;

  @include mq($from: md) {
    // Fallback for non flexbox browsers that support media queries e.g. IE9
    display: table;
    max-width: rem($nice-container-max-width);
    // TODO Replace with mixin from nice design system
    width: 96%;
    width: calc(100% - #{rem($spacing-large)});

    li {
      display: table-cell;
    }

    @supports (display: flex) {
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      justify-content: space-between;

      li {
        flex-grow: 1;
      }
    }
  }
}

.accountsMenuItem {
  @include mq($from: md) {
    display: none !important;
  }
}

.lastAccountsMenuItem {
  border-bottom: 1px solid $colour-nice-white;
  margin-bottom: rem($spacing-small);
  padding-bottom: rem($spacing-small);
}
