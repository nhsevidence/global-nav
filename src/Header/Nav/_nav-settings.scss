// TODO: Can we not use a magic number?
$sub-nav-height: 58;

$colour-nav-background: $colour-nice-teal;
$colour-nav-text: $colour-text-inverse;
