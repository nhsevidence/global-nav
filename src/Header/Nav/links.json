[
	{
		"href": "https://pathways.nice.org.uk/",
		"text": "NICE Pathways",
		"id": "pathways"
	},
	{
		"href": "https://www.nice.org.uk/guidance",
		"text": "NICE guidance"
	},
	{
		"href": "https://www.nice.org.uk/standards-and-indicators",
		"text": "Standards and indicators"
	},
	{
		"href": "https://www.evidence.nhs.uk/",
		"text": "Evidence search"
	},
	{
		"href": "https://bnf.nice.org.uk/",
		"text": "BNF",
		"abbreviation": true,
		"title": "British National Formulary"
	},
	{
		"href": "https://bnfc.nice.org.uk/",
		"text": "BNFC",
		"abbreviation": true,
		"title": "British National Formulary for Children"
	},
	{
		"href": "https://cks.nice.org.uk/",
		"text": "CKS",
		"abbreviation": true,
		"title": "Clinical Knowledge Summaries"
	},
	{
		"href": "https://www.nice.org.uk/about/what-we-do/evidence-services/journals-and-databases",
		"text": "Journals and databases"
	}
]
