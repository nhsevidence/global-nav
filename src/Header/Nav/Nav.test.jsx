import React from "react";
import Nav from "./Nav";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

jest.mock("./links.json", () => require("./__mocks__/links.json"));

describe("Nav", () => {
	const defaultProps = {
		isExpanded: false,
		accountsData: null,
		service: null
	};

	it("Renders without crashing", () => {
		const wrapper = shallow(<Nav {...defaultProps} />);
		expect(wrapper).toHaveLength(1);
	});

	it("Matches snapshot with no accounts links", () => {
		const wrapper = shallow(<Nav {...defaultProps} />);
		expect(toJson(wrapper)).toMatchSnapshot();
	});

	it("Matches snapshot with accounts links", () => {
		const accountsLinks = {
			"Joe Bloggs": "https://accounts.nice.org.uk/profile",
			"Sign out": "https://accounts.nice.org.uk/signout"
		};
		const wrapper = shallow(
			<Nav {...defaultProps} accountsLinks={accountsLinks} />
		);
		expect(toJson(wrapper.find("ul"))).toMatchSnapshot();
	});

	it("Adds aria current atrribute for selected service", () => {
		const wrapper = shallow(<Nav {...defaultProps} service="link2" />);
		expect(
			wrapper
				.find("a")
				.at(1)
				.props()["aria-current"]
		).toEqual(true);
	});

	it("Adds lastAccountsMenuItem css class to the last Accounts menu item", () => {
		const accountsLinks = {
			"Joe Bloggs": "https://accounts.nice.org.uk/profile",
			"Sign out": "https://accounts.nice.org.uk/signout"
		};
		const wrapper = shallow(
			<Nav {...defaultProps} accountsLinks={accountsLinks} />
		);
		expect(
			wrapper
				.find("li")
				.at(1)
				.props().className
		).toContain("lastAccountsMenuItem");
	});

	it("Add expanded class to nav when passed isExpanded prop", () => {
		const wrapper = shallow(<Nav {...defaultProps} isExpanded={false} />);
		expect(wrapper.props().className).not.toContain("navExpanded");
		wrapper.setProps({ isExpanded: true });
		expect(wrapper.props().className).toContain("navExpanded");
	});
});
