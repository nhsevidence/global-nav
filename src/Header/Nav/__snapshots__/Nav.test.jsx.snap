// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Matches snapshot with multiple accounts links 1`] = `
<nav
  aria-label="My account"
  className="nav myAccount"
>
  <h2
    className="myAccountHeading"
  >
    My account
  </h2>
  <div
    className="menuWrapper"
  >
    <ul
      className="menuList"
      role="menu"
    >
      <li
        key="https://accounts.nice.org.uk/profile"
        role="presentation"
      >
        <a
          className="link"
          href="https://accounts.nice.org.uk/profile"
          role="menuitem"
        >
          Joe Bloggs
        </a>
      </li>
      <li
        key="https://accounts.nice.org.uk/signout"
        role="presentation"
      >
        <a
          className="link"
          href="https://accounts.nice.org.uk/signout"
          role="menuitem"
        >
          Sign out
        </a>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`Nav Matches snapshot with no accounts links 1`] = `
<div
  className="wrapper"
  id="header-menu"
>
  <nav
    className="nav"
  >
    <div
      className="menuWrapper"
    >
      <ul
        aria-labelledby="header-menu-button"
        className="menuList"
        role="menu"
      >
        <li
          key="link1"
          role="presentation"
        >
          <a
            aria-current={null}
            className="link"
            href="https://url1/"
            role="menuitem"
          >
            <span>
              First link
            </span>
          </a>
        </li>
        <li
          key="link2"
          role="presentation"
        >
          <a
            aria-current={null}
            className="link"
            href="https://url2/"
            role="menuitem"
          >
            <span>
              <abbr
                title="Abbreviation title"
              >
                Second link
              </abbr>
            </span>
          </a>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;

exports[`Nav Matches snapshot with single accounts link 1`] = `
<nav
  aria-label="My account"
  className="nav myAccount"
>
  <div
    className="menuWrapper"
  >
    <ul
      className="menuList"
      role="menu"
    >
      <li
        key="https://accounts.nice.org.uk/"
        role="presentation"
      >
        <a
          className="link"
          href="https://accounts.nice.org.uk/"
          role="menuitem"
        >
          Sign in
        </a>
      </li>
    </ul>
  </div>
</nav>
`;

exports[`Nav Matches snapshot with sub links for selected service 1`] = `
<Nav
  accountsData={null}
  isExpanded={false}
  service="link2"
>
  <div
    className="wrapper wrapperWithSubLinks"
    id="header-menu"
  >
    <nav
      className="nav"
    >
      <div
        className="menuWrapper"
      >
        <ul
          aria-labelledby="header-menu-button"
          className="menuList"
          role="menu"
        >
          <li
            key="link1"
            role="presentation"
          >
            <a
              aria-current={null}
              className="link"
              href="https://url1/"
              role="menuitem"
            >
              <span>
                First link
              </span>
            </a>
          </li>
          <li
            key="link2"
            role="presentation"
          >
            <a
              aria-current={true}
              className="link"
              href="https://url2/"
              role="menuitem"
            >
              <span>
                <abbr
                  title="Abbreviation title"
                >
                  Second link
                </abbr>
              </span>
            </a>
            <SubNav
              links={
                Array [
                  Object {
                    "href": "/",
                    "text": "Sub link 1",
                  },
                  Object {
                    "href": "/test",
                    "text": "Sub link 2",
                  },
                ]
              }
              text="Second link"
            >
              <div
                className="wrapper"
              >
                <ul
                  aria-label="Second link links"
                  className="list"
                >
                  <li
                    key="0"
                  >
                    <a
                      aria-current="page"
                      className="link"
                      href="/"
                      role="menuitem"
                    >
                      Sub link 1
                    </a>
                  </li>
                  <li
                    key="1"
                  >
                    <a
                      aria-current={null}
                      className="link"
                      href="/test"
                      role="menuitem"
                    >
                      Sub link 2
                    </a>
                  </li>
                </ul>
              </div>
            </SubNav>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</Nav>
`;
